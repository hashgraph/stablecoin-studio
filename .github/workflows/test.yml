name: test
on:
  pull_request:
    branches: [ main, release/**, feature/** ]
  push:
    branches: [ main, release/** ]
    tags: [ v* ]
jobs:
  test-node:
    name: Unit Tests
    runs-on: ubuntu-latest
    steps:
      - name: Use Node.js TLS 16.x
        uses: actions/setup-node@v1
        with:
          node-version: 16.x
      - name: Checkout repository
        uses: actions/checkout@v2
      - name: Install dependencies
        run: npm i
      - name: Build
        run: npm run build
      - name: Test
        run: npm run test