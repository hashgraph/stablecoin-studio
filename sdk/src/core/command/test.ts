/* eslint-disable @typescript-eslint/no-empty-interface */

interface CommandResponse {}
interface ICommand {}

export class Command<T> implements ICommand {
	private $resultType!: T;
}

export interface ICommandBus {
	execute<X>(command: Command<X>): Promise<X>;
}

export type IInferringCommandHandler<CommandType extends Command<unknown>> =
	CommandType extends Command<infer ResultType>
		? {
				execute(command: CommandType): Promise<ResultType>;
		  }
		: never;

class ConcreteCommandResponse implements CommandResponse {
	constructor(public readonly status: string) {}
}

class ConcreteCommand extends Command<ConcreteCommandResponse> {
	constructor(public readonly value: string) {
		super();
	}
}

class CommandHandler<CommandType extends Command<unknown>>
	implements IInferringCommandHandler<ConcreteCommand>
{
	execute(command: ConcreteCommand): Promise<ConcreteCommandResponse> {
		throw new Error('Method not implemented.');
	}
}

class CommandBus implements ICommandBus {
	execute<X>(command: Command<X>): Promise<X> {
		throw new Error('Method not implemented.');
	}
}

// Test

const bus = new CommandBus();
const res = bus.execute(new ConcreteCommand('1'));
