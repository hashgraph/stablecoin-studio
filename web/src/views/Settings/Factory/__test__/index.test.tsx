import FactorySettings from '../';
import { render } from '../../../../test/index';
import translations from '../../../../translations/en/settings.json';
import userEvent from '@testing-library/user-event';
import configureMockStore from 'redux-mock-store';
import { Network } from '@hashgraph/stablecoin-npm-sdk';

const mockStore = configureMockStore();

describe(`<${FactorySettings.name} />`, () => {
	beforeEach(() => {});

	test('should render correctly', () => {
		const component = render(<FactorySettings />);

		expect(component.asFragment()).toMatchSnapshot();
	});

	test('should has title', () => {
		const component = render(<FactorySettings />);
		const header = component.getByTestId('base-container-heading');

		expect(header).toHaveTextContent(translations.title);
	});

	test('should have update owner and update implementation factory buttons', async () => {
		const store = mockStore({
			wallet: {
				isFactoryProxyOwner: true,
				isFactoryAcceptOwner: false,
				isFactoryPendingOwner: false,
			},
		});

		jest.mock('react-hook-form', () => ({
			...jest.requireActual('react-hook-form'),
			Controller: () => <></>,
			useForm: () => ({
				getValues: () => ({
					updateImplementation: '0.0.12345',
					updateOwner: '0.0.12345',
				}),
			}),
		}));

		jest.spyOn(Network, 'getFactoryAddress').mockReturnValue('0.0.12345');

		const component = render(<FactorySettings />, store);

		const stableCoin = await component.findByTestId('address-label');
		expect(stableCoin).toBeInTheDocument();
		expect(stableCoin).toBeEnabled();

		const addressButton = await component.findByTestId('update-implementation-address-button');
		expect(addressButton).toBeInTheDocument();
		expect(addressButton).toBeEnabled();
		await userEvent.click(addressButton);

		const ownerButton = await component.findByTestId('update-owner-button');
		expect(ownerButton).toBeInTheDocument();
		expect(ownerButton).toBeEnabled();
		await userEvent.click(ownerButton);
	});

	test('should have accept factory owner button', async () => {
		const store = mockStore({
			wallet: {
				isFactoryProxyOwner: false,
				isFactoryAcceptOwner: true,
				isFactoryPendingOwner: false,
			},
		});

		jest.mock('react-hook-form', () => ({
			...jest.requireActual('react-hook-form'),
			Controller: () => <></>,
			useForm: () => ({
				getValues: () => ({
					updateImplementation: '0.0.12345',
					updateOwner: '0.0.12345',
				}),
			}),
		}));

		jest.spyOn(Network, 'getFactoryAddress').mockReturnValue('0.0.12345');

		const component = render(<FactorySettings />, store);

		const acceptFactoryOwnerButton = await component.findByTestId('accept-factory-owner-button');
		expect(acceptFactoryOwnerButton).toBeInTheDocument();
		expect(acceptFactoryOwnerButton).toBeEnabled();
		await userEvent.click(acceptFactoryOwnerButton);
	});

	test('should have pending factory proxy owner buttons', async () => {
		const store = mockStore({
			wallet: {
				isFactoryProxyOwner: true,
				isFactoryAcceptOwner: false,
				isFactoryPendingOwner: true,
			},
		});

		jest.mock('react-hook-form', () => ({
			...jest.requireActual('react-hook-form'),
			Controller: () => <></>,
			useForm: () => ({
				getValues: () => ({
					updateImplementation: '0.0.12345',
					updateOwner: '0.0.12345',
				}),
			}),
		}));

		jest.spyOn(Network, 'getFactoryAddress').mockReturnValue('0.0.12345');

		const component = render(<FactorySettings />, store);

		const stableCoin = await component.findByTestId('address-label');
		expect(stableCoin).toBeInTheDocument();
		expect(stableCoin).toBeEnabled();

		const addressButton = await component.findByTestId('update-implementation-address-button');
		expect(addressButton).toBeInTheDocument();
		expect(addressButton).toBeEnabled();
		await userEvent.click(addressButton);

		const ownerButton = await component.findByTestId('update-owner-button');
		expect(ownerButton).toBeInTheDocument();
		expect(ownerButton).toBeEnabled();
		await userEvent.click(ownerButton);

		const cancelFactoryOwnerButton = await component.findByTestId('cancel-factory-owner-button');
		expect(cancelFactoryOwnerButton).toBeInTheDocument();
		expect(cancelFactoryOwnerButton).toBeEnabled();
		await userEvent.click(cancelFactoryOwnerButton);
	});
});
