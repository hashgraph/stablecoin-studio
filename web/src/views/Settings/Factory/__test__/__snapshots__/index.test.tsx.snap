// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<FactorySettings /> should render correctly 1`] = `
<DocumentFragment>
  <div
    class="chakra-stack css-16oka5y"
  >
    <h2
      class="chakra-heading css-bqsusc"
      data-testid="base-container-heading"
    >
      Settings
    </h2>
    <div
      class="css-1b6lda7"
      data-testid="base-container-children"
    />
  </div>
  <span
    hidden=""
    id="__chakra_env"
  />
</DocumentFragment>
`;
